domain/camera/src/main/java/com/google/jetpackcamera/domain/camera/EmptySurfaceProcessor.kt

```diff
@@ -0,0 +1,132 @@
+import androidx.camera.core.processing.OpenGlRenderer
+import androidx.camera.core.processing.ShaderProvider
+import androidx.core.util.Preconditions.checkState
+import java.util.concurrent.Executor
+
+private const val GL_THREAD_NAME = "EmptySurfaceProcessor"
+
+@SuppressLint("RestrictedApi")
+class EmptySurfaceProcessor : SurfaceProcessor {
+
+    private val glThread : HandlerThread = HandlerThread(GL_THREAD_NAME)
+    private var glHandler: Handler
+    var glExecutor: Executor
+        private set
+
+    // Members below are only accessed on GL thread.
+    private val glRenderer: OpenGlRenderer = OpenGlRenderer()
+    private val outputSurfaces: MutableMap<SurfaceOutput, Surface> = mutableMapOf()
+    private val textureTransform: FloatArray = FloatArray(16)
+    private val surfaceTransform: FloatArray = FloatArray(16)
+    private var isReleased = false
+
+    init {
+        glThread.start()
+        glHandler = Handler(glThread.looper)
+        glExecutor = newHandlerExecutor(glHandler)
+        glExecutor.execute {
+            glRenderer.init(
+                DynamicRange.SDR,
+                ShaderProvider.DEFAULT
+            )
+        }
+    }
+    override fun onInputSurface(surfaceRequest: SurfaceRequest) {
+        checkGlThread()
+        if (isReleased) {
+            surfaceRequest.willNotProvideSurface()
+            return
+        }
+        val surfaceTexture = SurfaceTexture(glRenderer.textureName)
+        surfaceTexture.setDefaultBufferSize(
+            surfaceRequest.resolution.width, surfaceRequest.resolution.height
+        )
+        val surface = Surface(surfaceTexture)
+        surfaceRequest.provideSurface(surface, glExecutor) {
+            surfaceTexture.setOnFrameAvailableListener(null)
+            surfaceTexture.release()
+            surface.release()
+        }
+        surfaceTexture.setOnFrameAvailableListener({
+            checkGlThread()
+            if (!isReleased) {
+                surfaceTexture.updateTexImage()
+                surfaceTexture.getTransformMatrix(textureTransform)
+                outputSurfaces.forEach {(surfaceOutput, surface) ->
+                    run {
+                        surfaceOutput.updateTransformMatrix(surfaceTransform, textureTransform)
+                        glRenderer.render(surfaceTexture.timestamp, surfaceTransform, surface)
+                    }
+                }
+            }
+        }, glHandler)
+    }
+
+    override fun onOutputSurface(surfaceOutput: SurfaceOutput) {
+        checkGlThread()
+        if (isReleased) {
+            surfaceOutput.close()
+            return
+        }
+        val surface = surfaceOutput.getSurface(glExecutor) {
+            surfaceOutput.close()
+            outputSurfaces.remove(surfaceOutput)?.let { removedSurface ->
+                glRenderer.unregisterOutputSurface(removedSurface)
+            }
+        }
+        glRenderer.registerOutputSurface(surface)
+        outputSurfaces[surfaceOutput] = surface
```

Comment:
This might be more of a question for Xi but is it possible that `outputSurfaces` already has an entry present for `surfaceOutput`? And if it does then we're replacing the entry with this new `surface` and not closing the old surface. So leaking a surface?
I wonder if it makes sense to assert that `outputSurfaces[surfaceOutput] == null` first?

Reply:
Looking at this implementation
https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:camera/camera-core/src/main/java/androidx/camera/core/processing/SurfaceOutputImpl.java

It doesn't look like this will be a problem.